Feature/user change

hansmaan opened the pull request:
죄송합니다. 생각보다 잡다한 오류가 많아서 좀 걸렸습니다. Post 부분은 조속히 진행해서 곧 끝낼 수 있도록 하겠습니다.

API 문서에 명시된 Password, Nickname, Email 바꾸는 기능, 정보 조회하는 기능, 유저 삭제하는 기능 도입했습니다.

hansmaan committed:
update university list

hansmaan committed:
setting change

hansmaan committed:
validation setting

hansmaan committed:
user update & delete

hansmaan commented on everytime/university/views.py:
```diff
@@ -10,7 +10,7 @@ class UniversityList(APIView):
     def get(self, request):
         serializer = UniversityViewSerializer
-        return Response(serializer(University.objects.all(), many=True).data, status=status.HTTP_200_OK)
+        return Response({"university_list" : serializer(University.objects.all(), many=True).data}, status=status.HTTP_200_OK)
```
API 문서와 좀 다른 것 같아서, 일치하도록 수정했습니다.

hansmaan commented on everytime/user/models.py:
```diff
@@ -52,7 +52,8 @@ class User(AbstractBaseUser, PermissionsMixin):
     user_id = models.CharField(max_length=30, blank=False, db_index=True, unique=True)
     email = models.EmailField(max_length=100, unique=True)
-    admission_year = models.IntegerField(max_length=10)
+    admission_year = models.IntegerField()
```
IntegerField는 max_length 안 붙이는게 맞다고 하네요.
is_active 만들어서 탈퇴한 회원인지 아닌지 구별할 수 있도록 했습니다.
is_active가 False인 회원은 아이디 비밀번호 쳐도 로그인 안 되는 로직이 내부에 구현되어 있는 것 같아요.

esc5221 replied:
오 기본 user모델에 적용되어있나보네요 신기하네

hansmaan commented on everytime/user/serializers.py:
```diff
@@ -16,23 +18,57 @@ def jwt_token_of(user):
+            'is_active'
+        )
 
     def validate(self, data):
```
update 함수를 따로 구현할 필요가 보였는데, 일반 Serializer에서는 구현이 불가능하더라구요. 그래서 ModelSerializer로 바꾸면서 전반적인 구조를 수정했습니다.

hansmaan commented on everytime/user/views.py:
```diff
@@ -18,19 +22,20 @@ class UserSignUpView(APIView):
-
         try:
+            validate_password(request.data.get('password'), user=request.user)
             user, jwt_token = serializer.save()
```
password가 주어진 형식에 어긋나지 않는지 파악하는 기능입니다.

hansmaan commented on everytime/user/views.py:
```diff
@@ -47,15 +52,47 @@ def post(self, request):
             except Exception as v:
                 logger.debug(v)
-                return Response({'error': "login_error", 'detail': "token login 오류"}, status=status.HTTP_400_BAD_REQUEST)
+                return Response({'error': "field_error", 'detail': "이메일 또는 비밀번호가 잘못되었습니다."}, status=status.HTTP_400_BAD_REQUEST)
```
아이디 비밀번호가 틀려서 로그인 실패한 것도, token login 오류라고 뜨는 것 같더라구요. 그냥 모든 login 실패에 대해서 token login 오류라고 출력하는 것 같아서, API 문서에 주어진 로그인 실패할 시의 출력과 동일시하게 만들었습니다.
어차피 둘 다 로그인 실패이고, token에 의한건지 아이디 비밀번호가 틀려서 실패한 것인지 구분하는 것은 힘든 일인 것 같아서 통합했습니다.

esc5221 replied:
네네 저도 일단 대충 그렇게 적어놨는데, api를 구분하던가 해야 명확히 오류가 구분될것 같아요.

hansmaan commented on everytime/user/views.py:
```diff
@@ -47,15 +52,47 @@ def post(self, request):
+        data['is_active'] = False
+        if not word:
+            return Response({"error" : "wrong_password", "detail" : "비밀번호를 입력해주세요."}, status= status.HTTP_400_BAD_REQUEST)
+        boolean = authenticate(user_id=user.user_id, password=word)
```
회원탈퇴 시, 비밀번호를 요구하고 틀릴 경우 탈퇴가 안 되게 하는 기능입니다.

hansmaan commented on everytime/user/views.py:
```diff
@@ -82,8 +119,8 @@ def get(self, request, *args, **kwargs):
             return Response({'닉네임을 입력해주세요.'}, status = status.HTTP_400_BAD_REQUEST)
```
"True" 보다는 True 그대로 출력해주는 것이 더 나은 것 같아서 기존에 했던 것들 좀 수정했습니다.

hansmaan commented on everytime/user/views.py:
```diff
@@ -96,5 +133,82 @@ def get(self, request, *args, **kwargs):
+            return Response({"success" : False, "detail" : "이미 사용중인 이메일입니다."}, status= status.HTTP_400_BAD_REQUEST)
+        return Response({"success" : True}, status = status.HTTP_200_OK)
+
+    def list(self, request, pk=None):
```
Update View 3개 다 list 함수가 있는데, put 함수만 갖다 놓으면 ViewSet 정의가 안 이루어지는 것 같아요. 계속 오류가 발생하길래 그냥 임의로 list 함수 하나 만들어 놓았습니다. (기능적 사유 X)

esc5221 replied:
그렇군요. 저도 왜 필요한지 궁금해서 없애봤는데 뭔가 잘 안되네요 ㅋㅋ 그대로 뒀습니다.

hansmaan commented on everytime/user/views.py:
```diff
@@ -96,5 +133,82 @@ def get(self, request, *args, **kwargs):
+        try:
+            validate_password(data.get('password'), user=request.user)
+            user.set_password(data.get('password'))
+            user.save()
```
Password가 형식에 맞는지 확인하고, 암호화해서 넣습니다.

hansmaan commented on everytime/common/validation.py:
```diff
@@ -0,0 +1,42 @@
+                code='password_too_long',
+                params={'max_length' : self.max_length}
+            )
+
```
둘다 Password의 형식을 판단하는 Validator입니다.
위는 최대 길이를 판별하는 Validator, 아래는 숫자, 문자, 문자열 중 두 가지 이상을 포함하는지 판별하는 Validator입니다.

esc5221 replied:
max_length 는 20자로 설정하셨군요. 확인했습니다.

hansmaan commented on everytime/everytime/settings.py:
```diff
@@ -74,6 +74,21 @@ def get_secret(setting, secrets=secrets):
+        'rest_framework_jwt.authentication.JSONWebTokenAuthentication',
+    ),
+}
+
```
User my ~~의 경우 토큰이 제대로 주어진 경우에 한해서만 진행할 수 있도록 로직을 짰습니다. 로그인 안 된 경우에는 401 Unauthorized Error가 발생하게 될 것입니다.

hansmaan commented on everytime/everytime/settings.py:
```diff
@@ -125,7 +140,7 @@ def get_secret(setting, secrets=secrets):
             'USER': 'team11_db_user',
-            'PASSWORD': 'secretkey1234',
+            'PASSWORD': 'secret1234',
         }
```
아 로컬에서 쓰느라 수정 못했습니다...

esc5221 replied:
revision PR에 수정해뒀습니다.

hansmaan commented on everytime/everytime/settings.py:
```diff
@@ -134,18 +149,18 @@ def get_secret(setting, secrets=secrets):
-        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
+        'NAME': 'common.validation.MaximumLengthValidator'
     },
     {
```
Password 양식 판별하는데 사용한 Validator들입니다.

esc5221 committed:
refactor : PR#17_revision
- my/email/ 및 2개의 API가 바꿀 데이터를 받는 방법을  query_param에서
data로 변경
- View에서 data 받는 부분 간소화 (partial 이용)
- UserUpdate*View 데이터 가져오는 방식 query_param에서 body로 변경
(API에 따름)
- UserCreateSerializer.validate 함수 간소화
- settings.py local mysql 비밀번호 변경

esc5221 commented:
https://github.com/wafflestudio19-5/team11-server/pull/18 현재 PR 살짝 수정해보았는데, 이걸 현재 branch로 merge 시키고, 이후 master로 merge하겠습니다.

hansmaan committed:
Merge pull request #18 from wafflestudio19-5/refactor/PR#17_revision

refactor : PR#17_revision
